File: legacy/perl/topsites.pl
#!/usr/bin/perl -w


# Makes a request to the Alexa Top Sites
# to get a list of top sites in a particular country

use  strict;

# If you don't have all the modules below, you can get them from CPAN
#
# perl -MCPAN -e 'install URI::Escape'
# perl -MCPAN -e 'install XML::XPath'
# etc...

use Digest::SHA qw(hmac_sha256_base64), 'sha256_hex','hmac_sha256','hmac_sha256_hex';
use File::Basename;
use URI::Escape;
use LWP::UserAgent;

use XML::XPath;
use XML::XPath::XMLParser;

my $service_endpoint = "ats.amazonaws.com";
my $service_host = "ats.us-west-1.amazonaws.com";
my $action = "TopSites";
my $service_port = 443;
my $service_uri = "/api";
my $service_region = "us-west-1";
my $service_name = "AlexaTopSites";
my $timestamp = generate_timestamp();

my $access_key_id;
my $secret_access_key;
my $country_code;

if ($#ARGV < 1) {
    print STDERR "Usage: " . basename($0) .
        " ACCESS_KEY_ID SECRET_ACCESS_KEY [COUNTRY_CODE]\n";
    exit(-1);
}
else {
    $access_key_id = $ARGV[0];
    $secret_access_key = $ARGV[1];
    $country_code =  ($#ARGV > 1) ? $ARGV[2] : "";
}

my $start = 1;
my $count = 10; # Max value is 100
my ($request_url, $authorization_header) = compose_url($start, $count);

print "Making request to:\n$request_url\n\n";

my $user_agent = LWP::UserAgent->new;
$user_agent->timeout(3);

my $request = HTTP::Request->new('GET', $request_url);
$request->header( 'Accept' => 'application/xml' );
$request->header( 'Content-Type' => 'application/xml' );
$request->header( 'x-amz-date' => $timestamp );
$request->header( 'Authorization' => $authorization_header );
my $response = $user_agent->request($request);

# If valid response, get the contents of the URL,
#  otherwise print error and return empty string.

my $link = "";
if ($response->is_success) {
  print "Sites:\n\n";

  my $output = $response->content;

  # print $output;

  my $xp = XML::XPath->new(xml => $output);

  my @site_nodes = $xp->findnodes('/aws:TopSitesResponse/aws:Response/aws:TopSitesResult/aws:Alexa/aws:TopSites/aws:Country/aws:Sites/aws:Site/aws:DataUrl');
  foreach my $node (@site_nodes) {
      my $value = $node->findvalue('.');
      print "$value\n";
  }
} else {
    print STDERR "Error connecting to $request_url. HTTP Status code: ". $response->code . "\nResponse: \n" . $response->content;
}

sub signing_key {
    my ($kSecret,$service,$region,$date) = @_;
    my $kDate    = hmac_sha256($date,'AWS4'.$kSecret);
    my $kRegion  = hmac_sha256($region,$kDate);
    my $kService = hmac_sha256($service,$kRegion);
    my $kSigning = hmac_sha256('aws4_request',$kService);
    return $kSigning;
}

sub calculate_signature {
    my ($kSecret,$service,$region,$date,$string_to_sign) = @_;
    print "SERVICE: " . $service . "\n";
    my $kSigning = signing_key($kSecret,$service,$region,$date);
    return hmac_sha256_hex($string_to_sign,$kSigning);
}

# Returns the AWS URL to get the site list for the specified country
sub compose_url {
    my ($site) = @_;
    my $uri = {
        'Action'           => $action,
        'ResponseGroup'    => 'Country',
        'CountryCode'      => $country_code,
        'Count'            => $count,
        'Start'            => $start
    };
    my $headers = {
      "host"        => $service_host,
      "x-amz-date"  => $timestamp
    };

    my @hdrs_arr = ();
    my @hdrl_arr = ();
    foreach my $key (sort keys %$headers) {
        push(@hdrs_arr, escape($key) . ':' . escape($headers->{$key}));
        push(@hdrl_arr, escape($key));
    }
    my $hdr_str = join("\n", @hdrs_arr) . "\n";
    my $hdr_lst = join(';', @hdrl_arr);

    my @uri_arr = ();
    #sort hash and uri escape
    foreach my $key (sort keys %$uri) {
        push(@uri_arr, escape($key) . '=' . escape($uri->{$key}));
    }

    my $uri_str = join('&', @uri_arr);
    my $payload_hash = sha256_hex("");
    my $canonical_request = "GET" . "\n" . $service_uri . "\n" . $uri_str . "\n" . $hdr_str . "\n" . $hdr_lst . "\n" . $payload_hash;
    my $algorithm = "AWS4-HMAC-SHA256";
    my $credential_scope = substr($timestamp, 0, 8) . "/" . $service_region . "/" . $service_name . "/" . "aws4_request";
    my $string_to_sign = $algorithm . "\n" .  $timestamp . "\n" .  $credential_scope . "\n" . sha256_hex($canonical_request);
    my $signature =  calculate_signature($secret_access_key,$service_name,$service_region,substr($timestamp, 0, 8),$string_to_sign);
    my $authorization_header = $algorithm . " " . "Credential=" . $access_key_id . "/" . $credential_scope . ", " .  "SignedHeaders=" . $hdr_lst . ", " . "Signature=" . $signature;

    return ('https://' . $service_endpoint . $service_uri . '?' . $uri_str, $authorization_header);
}

# Calculate current TimeStamp
sub generate_timestamp {
    return sprintf("%04d%02d%02dT%02d%02d%02dZ",
                   sub { ($_[5]+1900,
                          $_[4]+1,
                          $_[3],
                          $_[2],
                          $_[1],
                          $_[0])
                   }->(gmtime(time)));
}


# URI escape only the characters that should be escaped, according to RFC 3986
sub escape {
    my ($str) = @_;
    return uri_escape_utf8($str,'^A-Za-z0-9\-_.~');
}

# The digest is the signature
sub digest {
    my ($query) = @_;
    my $digest = hmac_sha256_base64 ($query, $secret_access_key);

    # Digest::MMM modules do not pad their base64 output, so we do
    # it ourselves to keep the service happy.
    return $digest . "=";
}

File: legacy/php/topsites.php
<?php
/**
 * Makes a request to ATS for the top 10 sites in a country
 */
class TopSites {

    protected static $ActionName        = 'Topsites';
    protected static $ResponseGroupName = 'Country';
    protected static $ServiceEndpoint   = 'ats.us-west-1.amazonaws.com';
    protected static $ServiceHost       = 'ats.amazonaws.com';
    protected static $NumReturn         = 10;
    protected static $StartNum          = 1;
    protected static $SigVersion        = '2';
    protected static $HashAlgorithm     = 'HmacSHA256';
    protected static $ServiceURI = "/api";
    protected static $ServiceRegion = "us-west-1";
    protected static $ServiceName = "AlexaTopSites";


    public function TopSites($accessKeyId, $secretAccessKey, $countryCode) {
        $this->accessKeyId = $accessKeyId;
        $this->secretAccessKey = $secretAccessKey;
        $this->countryCode = $countryCode;
        $now = time();
        $this->amzDate = gmdate("Ymd\THis\Z", $now);
        $this->dateStamp = gmdate("Ymd", $now);

    }

    /**
     * Get site info from AWIS.
     */
    public function getTopSites() {
        $canonicalQuery = $this->buildQueryParams();
        $canonicalHeaders =  $this->buildHeaders(true);
        $signedHeaders = $this->buildHeaders(false);
        $payloadHash = hash('sha256', "");
        $canonicalRequest = "GET" . "\n" . self::$ServiceURI . "\n" . $canonicalQuery . "\n" . $canonicalHeaders . "\n" . $signedHeaders . "\n" . $payloadHash;
        $algorithm = "AWS4-HMAC-SHA256";
        $credentialScope = $this->dateStamp . "/" . self::$ServiceRegion . "/" . self::$ServiceName . "/" . "aws4_request";
        $stringToSign = $algorithm . "\n" .  $this->amzDate . "\n" .  $credentialScope . "\n" .  hash('sha256', $canonicalRequest);
        $signingKey = $this->getSignatureKey();
        $signature = hash_hmac('sha256', $stringToSign, $signingKey);
        $authorizationHeader = $algorithm . ' ' . 'Credential=' . $this->accessKeyId . '/' . $credentialScope . ', ' .  'SignedHeaders=' . $signedHeaders . ', ' . 'Signature=' . $signature;

        $url = 'https://' . self::$ServiceHost . self::$ServiceURI . '?' . $canonicalQuery;
        $ret = self::makeRequest($url, $authorizationHeader);
        self::parseResponse($ret);
    }

    protected function sign($key, $msg) {
        return hash_hmac('sha256', $msg, $key, true);
    }

    protected function getSignatureKey() {
        $kSecret = 'AWS4' . $this->secretAccessKey;
        $kDate = $this->sign($kSecret, $this->dateStamp);
        $kRegion = $this->sign($kDate, self::$ServiceRegion);
        $kService = $this->sign($kRegion, self::$ServiceName);
        $kSigning = $this->sign($kService, 'aws4_request');
        return $kSigning;
    }

    /**
     * Builds headers for the request to AWIS.
     * @return String headers for the request
     */
    protected function buildHeaders($list) {
        $params = array(
            'host'            => self::$ServiceEndpoint,
            'x-amz-date'      => $this->amzDate
        );
        ksort($params);
        $keyvalue = array();
        foreach($params as $k => $v) {
            if ($list)
              $keyvalue[] = $k . ':' . $v;
            else {
              $keyvalue[] = $k;
            }
        }
        return ($list) ? implode("\n",$keyvalue) . "\n" : implode(';',$keyvalue) ;
    }

    /**
     * Builds query parameters for the request to AWIS.
     * Parameter names will be in alphabetical order and
     * parameter values will be urlencoded per RFC 3986.
     * @return String query parameters for the request
     */
    protected function buildQueryParams() {
        $params = array(
          'Action'            => self::$ActionName,
          'ResponseGroup'     => self::$ResponseGroupName,
          'CountryCode'       => $this->countryCode,
          'Count'             => self::$NumReturn,
          'Start'             => self::$StartNum
        );
        ksort($params);
        $keyvalue = array();
        foreach($params as $k => $v) {
            $keyvalue[] = $k . '=' . rawurlencode($v);
        }
        return implode('&',$keyvalue);
    }

    /**
     * Makes request to TopSites
     * @param String $url   URL to make request to
     * @param String authorizationHeader  Authorization string
     * @return String       Result of request
     */
    protected function makeRequest($url, $authorizationHeader) {
        echo "\nMaking request to:\n$url\n";
        $ch = curl_init($url);
        curl_setopt($ch, CURLOPT_TIMEOUT, 4);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($ch, CURLOPT_HTTPHEADER, array(
          'Accept: application/xml',
          'Content-Type: application/xml',
          'X-Amz-Date: ' . $this->amzDate,
          'Authorization: ' . $authorizationHeader
        ));
        $result = curl_exec($ch);
        curl_close($ch);
        return $result;
    }

    /**
     * Parses the XML response from ATS and echoes the DataUrl element
     * for each returned site
     *
     * @param String $response    xml response from ATS
     */
    protected static function parseResponse($response) {
        echo "\nSites: \n";
        $xml = new SimpleXMLElement($response,null, false, 'aws', true);
        foreach($xml->Response->TopSitesResult->Alexa->TopSites->Country->Sites->children('aws', true) as $site) {
            echo $site->DataUrl . "\n";
        }
    }

}

if (count($argv) < 3) {
    echo "Usage: $argv[0] ACCESS_KEY_ID SECRET_ACCESS_KEY [COUNTRY_CODE]\n";
    exit(-1);
}
else {
    $accessKeyId = $argv[1];
    $secretAccessKey = $argv[2];
    $countryCode = count($argv) > 3 ? $argv[3] : "";
}

$topSites = new TopSites($accessKeyId, $secretAccessKey, $countryCode);
$topSites->getTopSites();

?>
